File: Планшет/app/src/main/java/ua/church/slovo/remote/SermonPlan.java
package ua.church.slovo.remote;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

/// План проповіді: назва й пункти в порядку подачі.
///
/// Кожен план — окремий файл у пам'яті планшета (`files/plans/*.json`):
/// проповідник може скласти кілька наперед і взяти на служіння потрібний.
/// Пункт везе з собою не лише посилання, а й сам текст — вірші чи слова
/// пісні. Якщо в програмі на служінні не виявиться того перекладу чи
/// пісенника, пункт усе одно покажеться — текстом.
final class SermonPlan {

    static final String SCRIPTURE = "scripture";
    static final String SONG = "song";
    static final String TEXT = "text";
    static final String FILE = "file";

    static final class Item {
        String type = TEXT;
        /// Рядок у списку: «Ів 3:16-18», назва пісні, заголовок, ім'я файла.
        String title = "";

        // Уривок.
        String module = "";
        String moduleName = "";
        String bookName = "";
        int canon;
        int book;
        int chapter;
        final List<Integer> verses = new ArrayList<>();
        String text = "";

        // Пісня.
        String songBook = "";
        int song;
        final List<String[]> parts = new ArrayList<>();

        // Довільний текст.
        String heading = "";
        String body = "";

        // Файл: копія в пам'яті планшета і його справжнє ім'я.
        String path = "";
        String name = "";

        /// Друга половина рядка списку — початок тексту.
        String subtitle() {
            String value;
            switch (type) {
                case SCRIPTURE: value = text; break;
                case SONG: value = parts.isEmpty() ? "" : parts.get(0)[1]; break;
                case TEXT: value = body; break;
                default: value = name; break;
            }
            value = value.replace('\n', ' ').trim();
            return value.length() > 120 ? value.substring(0, 120) + "…" : value;
        }

        JSONObject toJson() throws JSONException {
            JSONObject json = new JSONObject();
            json.put("type", type);
            json.put("title", title);
            switch (type) {
                case SCRIPTURE:
                    json.put("module", module);
                    json.put("moduleName", moduleName);
                    json.put("bookName", bookName);
                    json.put("canon", canon);
                    json.put("book", book);
                    json.put("chapter", chapter);
                    JSONArray numbers = new JSONArray();
                    for (Integer verse : verses) numbers.put(verse);
                    json.put("verses", numbers);
                    json.put("text", text);
                    break;
                case SONG:
                    json.put("module", module);
                    json.put("songBook", songBook);
                    json.put("song", song);
                    JSONArray list = new JSONArray();
                    for (String[] part : parts) {
                        JSONObject item = new JSONObject();
                        item.put("kind", part[0]);
                        item.put("text", part[1]);
                        list.put(item);
                    }
                    json.put("parts", list);
                    break;
                case TEXT:
                    json.put("heading", heading);
                    json.put("body", body);
                    break;
                default:
                    json.put("path", path);
                    json.put("name", name);
                    break;
            }
            return json;
        }

        static Item fromJson(JSONObject json) {
            Item item = new Item();
            item.type = json.optString("type", TEXT);
            item.title = json.optString("title", "");
            item.module = json.optString("module", "");
            item.moduleName = json.optString("moduleName", "");
            item.bookName = json.optString("bookName", "");
            item.canon = json.optInt("canon");
            item.book = json.optInt("book");
            item.chapter = json.optInt("chapter");
            JSONArray numbers = json.optJSONArray("verses");
            if (numbers != null) for (int i = 0; i < numbers.length(); i++) item.verses.add(numbers.optInt(i));
            item.text = json.optString("text", "");
            item.songBook = json.optString("songBook", "");
            item.song = json.optInt("song");
            JSONArray list = json.optJSONArray("parts");
            if (list != null) {
                for (int i = 0; i < list.length(); i++) {
                    JSONObject part = list.optJSONObject(i);
                    if (part != null) item.parts.add(new String[] { part.optString("kind"), part.optString("text") });
                }
            }
            item.heading = json.optString("heading", "");
            item.body = json.optString("body", "");
            item.path = json.optString("path", "");
            item.name = json.optString("name", "");
            return item;
        }
    }

    String id = UUID.randomUUID().toString();
    String title = "";
    long updated;
    final List<Item> items = new ArrayList<>();

    static File folder(Context context) {
        File folder = new File(context.getFilesDir(), "plans");
        //noinspection ResultOfMethodCallIgnored
        folder.mkdirs();
        return folder;
    }

    /// Копії файлів плану: презентацію мають відвезти й тоді, коли вихідний
    /// файл на планшеті вже прибрали.
    static File filesFolder(Context context) {
        File folder = new File(context.getFilesDir(), "plan-files");
        //noinspection ResultOfMethodCallIgnored
        folder.mkdirs();
        return folder;
    }

    /// Усі плани, свіжіші зверху.
    static List<SermonPlan> all(Context context) {
        List<SermonPlan> result = new ArrayList<>();
        File[] files = folder(context).listFiles();
        if (files != null) {
            for (File file : files) {
                if (!file.getName().endsWith(".json")) continue;
                SermonPlan plan = load(file);
                if (plan != null) result.add(plan);
            }
        }
        Collections.sort(result, (a, b) -> Long.compare(b.updated, a.updated));
        return result;
    }

    static SermonPlan load(File file) {
        try {
            JSONObject json = new JSONObject(new String(ModuleReaders.readAll(file), StandardCharsets.UTF_8));
            SermonPlan plan = new SermonPlan();
            plan.id = json.optString("id", file.getName().replace(".json", ""));
            plan.title = json.optString("title", "");
            plan.updated = json.optLong("updated", file.lastModified());
            JSONArray items = json.optJSONArray("items");
            if (items != null) {
                for (int i = 0; i < items.length(); i++) {
                    JSONObject item = items.optJSONObject(i);
                    if (item != null) plan.items.add(Item.fromJson(item));
                }
            }
            return plan;
        } catch (IOException | JSONException error) {
            return null;
        }
    }

    static SermonPlan load(Context context, String id) {
        return id == null || id.isEmpty() ? null : load(new File(folder(context), id + ".json"));
    }

    void save(Context context) throws IOException {
        updated = System.currentTimeMillis();
        JSONObject json = new JSONObject();
        try {
            json.put("id", id);
            json.put("title", title);
            json.put("updated", updated);
            JSONArray list = new JSONArray();
            for (Item item : items) list.put(item.toJson());
            json.put("items", list);
        } catch (JSONException error) {
            throw new IOException(error.getMessage());
        }
        File target = new File(folder(context), id + ".json");
        File temp = new File(folder(context), id + ".tmp");
        try (OutputStream out = new FileOutputStream(temp)) {
            out.write(json.toString().getBytes(StandardCharsets.UTF_8));
        }
        if (!temp.renameTo(target)) throw new IOException(Lang.t("Не вдалося записати план", "Could not save the plan"));
    }

    /// Прибрати план разом із копіями його файлів.
    void delete(Context context) {
        for (Item item : items) {
            if (FILE.equals(item.type) && !item.path.isEmpty()) {
                //noinspection ResultOfMethodCallIgnored
                new File(item.path).delete();
            }
        }
        //noinspection ResultOfMethodCallIgnored
        new File(folder(context), id + ".json").delete();
    }
}

File: Проповідник/Sources/SermonPlan.cs
// =============================================================================
//  SermonPlan.cs — план проповіді: назва й пункти в порядку подачі
// =============================================================================
//  Перенесено з планшета (SermonPlan.java), формат файла той самий.
//  Кожен план — окремий файл у теці plans: проповідник може скласти кілька
//  наперед і взяти на служіння потрібний. Пункт везе з собою не лише
//  посилання, а й сам текст — вірші чи слова пісні: якщо в програмі на
//  служінні не виявиться того перекладу чи пісенника, пункт усе одно
//  покажеться — текстом.
// =============================================================================

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Propovidnyk;

public sealed class PlanItem
{
    public const string Scripture = "scripture";
    public const string Song = "song";
    public const string Text = "text";
    public const string File = "file";

    public string Type = Text;
    /// Рядок у списку: «Ів 3:16-18», назва пісні, заголовок, ім'я файла.
    public string Title = "";

    // Уривок.
    public string Module = "";
    public string ModuleName = "";
    public string BookName = "";
    public int Canon;
    public int Book;
    public int Chapter;
    public List<int> Verses = new();
    public string Body = "";   // текст віршів (для уривка) — у файлі поле "text"

    // Пісня.
    public string SongBook = "";
    public int SongIndex;
    public List<(string Kind, string Text)> Parts = new();

    // Довільний текст.
    public string Heading = "";
    public string Announcement = "";   // у файлі поле "body"

    // Файл: копія в теці програми і його справжнє ім'я.
    public string Path = "";
    public string Name = "";

    public string Icon => Type switch
    {
        Scripture => "📖",
        Song => "🎵",
        File => "📄",
        _ => "✎",
    };

    /// Друга половина рядка списку — початок тексту.
    public string Subtitle
    {
        get
        {
            var value = Type switch
            {
                Scripture => Body,
                Song => Parts.Count == 0 ? "" : Parts[0].Text,
                Text => Announcement,
                _ => Name,
            };
            value = value.Replace('\r', ' ').Replace('\n', ' ').Trim();
            return value.Length > 120 ? value[..120] + "…" : value;
        }
    }

    public JsonObject ToJson()
    {
        var json = new JsonObject { ["type"] = Type, ["title"] = Title };
        switch (Type)
        {
            case Scripture:
                json["module"] = Module;
                json["moduleName"] = ModuleName;
                json["bookName"] = BookName;
                json["canon"] = Canon;
                json["book"] = Book;
                json["chapter"] = Chapter;
                json["verses"] = new JsonArray(Verses.Select(v => (JsonNode)v).ToArray());
                json["text"] = Body;
                break;
            case Song:
                json["module"] = Module;
                json["songBook"] = SongBook;
                json["song"] = SongIndex;
                json["parts"] = new JsonArray(Parts.Select(p => (JsonNode)new JsonObject { ["kind"] = p.Kind, ["text"] = p.Text }).ToArray());
                break;
            case Text:
                json["heading"] = Heading;
                json["body"] = Announcement;
                break;
            default:
                json["path"] = Path;
                json["name"] = Name;
                break;
        }
        return json;
    }

    public static PlanItem FromJson(JsonObject json)
    {
        var item = new PlanItem
        {
            Type = (string?)json["type"] ?? Text,
            Title = (string?)json["title"] ?? "",
            Module = (string?)json["module"] ?? "",
            ModuleName = (string?)json["moduleName"] ?? "",
            BookName = (string?)json["bookName"] ?? "",
            Canon = (int?)json["canon"] ?? 0,
            Book = (int?)json["book"] ?? 0,
            Chapter = (int?)json["chapter"] ?? 0,
            SongBook = (string?)json["songBook"] ?? "",
            SongIndex = (int?)json["song"] ?? 0,
            Heading = (string?)json["heading"] ?? "",
            Path = (string?)json["path"] ?? "",
            Name = (string?)json["name"] ?? "",
        };
        if (json["verses"] is JsonArray verses) item.Verses = verses.Select(v => (int?)v ?? 0).ToList();
        if (json["parts"] is JsonArray parts)
            item.Parts = parts.OfType<JsonObject>().Select(p => ((string?)p["kind"] ?? "", (string?)p["text"] ?? "")).ToList();
        // «text» — текст віршів, «body» — текст оголошення.
        item.Body = item.Type == Scripture ? (string?)json["text"] ?? "" : "";
        item.Announcement = item.Type == Text ? (string?)json["body"] ?? "" : "";
        return item;
    }
}

public sealed class SermonPlan
{
    public string Id = Guid.NewGuid().ToString();
    public string Title = "";
    public long Updated;
    public List<PlanItem> Items = new();

    public string DisplayTitle => string.IsNullOrWhiteSpace(Title) ? Lang.T("План без назви", "Untitled plan") : Title;

    /// Усі плани, свіжіші зверху.
    public static List<SermonPlan> All()
    {
        var result = new List<SermonPlan>();
        foreach (var file in Directory.GetFiles(Paths.Plans, "*.json"))
        {
            var plan = Load(file);
            if (plan != null) result.Add(plan);
        }
        return result.OrderByDescending(p => p.Updated).ToList();
    }

    public static SermonPlan? Load(string file)
    {
        try
        {
            if (JsonNode.Parse(System.IO.File.ReadAllText(file)) is not JsonObject json) return null;
            var plan = new SermonPlan
            {
                Id = (string?)json["id"] ?? System.IO.Path.GetFileNameWithoutExtension(file),
                Title = (string?)json["title"] ?? "",
                Updated = (long?)json["updated"] ?? new DateTimeOffset(System.IO.File.GetLastWriteTimeUtc(file)).ToUnixTimeMilliseconds(),
            };
            if (json["items"] is JsonArray items)
                plan.Items = items.OfType<JsonObject>().Select(PlanItem.FromJson).ToList();
            return plan;
        }
        catch (Exception error)
        {
            Paths.Say($"план {file} не прочитався: {error.Message}");
            return null;
        }
    }

    public static SermonPlan? Load(string? id, bool byId) =>
        string.IsNullOrEmpty(id) ? null : Load(System.IO.Path.Combine(Paths.Plans, id + ".json"));

    public void Save()
    {
        Updated = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var json = new JsonObject
        {
            ["id"] = Id,
            ["title"] = Title,
            ["updated"] = Updated,
            ["items"] = new JsonArray(Items.Select(i => (JsonNode)i.ToJson()).ToArray()),
        };
        var target = System.IO.Path.Combine(Paths.Plans, Id + ".json");
        var temp = System.IO.Path.Combine(Paths.Plans, Id + ".tmp");
        System.IO.File.WriteAllText(temp, json.ToJsonString(new JsonSerializerOptions { WriteIndented = true }));
        System.IO.File.Move(temp, target, true);
    }

    /// Прибрати план разом із копіями його файлів.
    public void Delete()
    {
        foreach (var item in Items)
        {
            if (item.Type == PlanItem.File && item.Path.Length > 0)
            {
                try { System.IO.File.Delete(item.Path); } catch { /* уже нема */ }
            }
        }
        try { System.IO.File.Delete(System.IO.Path.Combine(Paths.Plans, Id + ".json")); } catch { /* уже нема */ }
    }

    /// Вірші, що йдуть підряд, згортаються в діапазон: 1,2,3,7 → «1-3,7».
    public static string Span(IReadOnlyList<int> numbers)
    {
        var parts = new List<string>();
        var i = 0;
        while (i < numbers.Count)
        {
            var start = numbers[i];
            var end = start;
            while (i + 1 < numbers.Count && numbers[i + 1] == end + 1)
            {
                i++;
                end = numbers[i];
            }
            parts.Add(start == end ? start.ToString() : start + "-" + end);
            i++;
        }
        return string.Join(",", parts);
    }
}
